src/MemoryTracing/MemoryTracingIntegrationTest.cpp

```diff
@@ -131,120 +186,77 @@ class MemoryTracingIntegrationTestFixture {
+      EXPECT_TRUE(cgroup_memory_usage.unevictable_bytes() >= 0);
+      EXPECT_TRUE(cgroup_memory_usage.inactive_anon_bytes() >= 0);
+      EXPECT_TRUE(cgroup_memory_usage.active_anon_bytes() >= 0);
+      EXPECT_TRUE(cgroup_memory_usage.inactive_file_bytes() >= 0);
+      EXPECT_TRUE(cgroup_memory_usage.active_file_bytes() >= 0);
     }
+
+    // Verify that the memory events in the same wrapper are sampled at very close time.
+    uint64_t max_timestamp = *std::max_element(timestamps.begin(), timestamps.end());
+    uint64_t min_timestamp = *std::min_element(timestamps.begin(), timestamps.end());
+    EXPECT_LE(max_timestamp - min_timestamp, kMemoryEventsTimeDifferenceTolerace);
   }
 }
 
 // Verify whether the memory_sampling_period_ns_ works as expected by checking the number of
 // received events.
-void VerifyEventCounts(const std::vector<ProducerCaptureEvent>& events,
-                       size_t num_expected_events_per_type) {
+void VerifyEventCounts(const std::vector<ProducerCaptureEvent>& events, size_t expected_counts) {
   constexpr size_t kEventCountsErrorTolerance = 2;
 
-  size_t num_received_system_memory_events = 0;
-  size_t num_received_cgroup_memory_events = 0;
-  size_t num_received_process_memory_events = 0;
+  size_t num_received_wrappers = 0;
   for (const auto& event : events) {
-    switch (event.event_case()) {
-      case ProducerCaptureEvent::kSystemMemoryUsage: {
-        num_received_system_memory_events++;
-        break;
-      }
-      case ProducerCaptureEvent::kCgroupMemoryUsage: {
-        num_received_cgroup_memory_events++;
-        break;
-      }
-      case ProducerCaptureEvent::kProcessMemoryUsage: {
-        num_received_process_memory_events++;
-        break;
-      }
-      default:
-        UNREACHABLE();
-    }
+    if (event.event_case() == ProducerCaptureEvent::kMemoryEventWrapper) num_received_wrappers++;
   }
 
-  EXPECT_GE(num_received_system_memory_events,
-            num_expected_events_per_type - kEventCountsErrorTolerance);
-  EXPECT_LE(num_received_system_memory_events,
-            num_expected_events_per_type + kEventCountsErrorTolerance);
-
-  EXPECT_GE(num_received_process_memory_events,
-            num_expected_events_per_type - kEventCountsErrorTolerance);
-  EXPECT_LE(num_received_process_memory_events,
-            num_expected_events_per_type + kEventCountsErrorTolerance);
-
-  // Verify the number of cgroup memory events only if the process's memory cgroup and the cgroup
-  // memory.stat file can be found successfully
-  if (num_received_cgroup_memory_events == 0) return;
-  EXPECT_GE(num_received_cgroup_memory_events,
-            num_expected_events_per_type - kEventCountsErrorTolerance);
-  EXPECT_LE(num_received_cgroup_memory_events,
-            num_expected_events_per_type + kEventCountsErrorTolerance);
+  EXPECT_GE(num_received_wrappers, expected_counts - kEventCountsErrorTolerance);
+  EXPECT_LE(num_received_wrappers, expected_counts + kEventCountsErrorTolerance);
```

Comment:
I could expect to receive less but how we could receive more events?

Reply:
Ah, right, the `EXPECT_LE` case is redundant. Removed now. Thanks!